File: typescript/src/api/github.ts
import { Octokit } from '@octokit/rest';
import { format } from 'date-fns';
import { utcToZonedTime } from 'date-fns-tz';
import { Config } from '../config';

export interface WorklogEntry {
  content: string;
  startIndex: number;
  endIndex: number;
}

export class GitHubApiClient {
  private readonly octokit: Octokit;
  private readonly config: Config;
  
  constructor(config: Config) {
    this.config = config;
    this.octokit = new Octokit({
      auth: config.githubToken,
    });
  }
  
  /**
   * Get the current content of the worklog file
   * @returns Tuple of [content, sha]
   */
  async getWorklogContent(): Promise<[string, string]> {
    const { githubRepo, githubWorklogPath, logger } = this.config;
    
    // Split repo into owner and repo
    const [owner, repo] = githubRepo.split('/');
    
    if (!owner || !repo) {
      throw new Error(`Invalid GitHub repository format: ${githubRepo}. Expected format: owner/repo`);
    }
    
    logger.info(`Fetching worklog content from ${githubRepo}/${githubWorklogPath}`);
    
    try {
      const response = await this.octokit.repos.getContent({
        owner,
        repo,
        path: githubWorklogPath,
      });
      
      // Check if response is a file (not a directory)
      if (Array.isArray(response.data)) {
        throw new Error(`Expected a file, but ${githubWorklogPath} is a directory`);
      }
      
      if (!('content' in response.data) || !('sha' in response.data)) {
        throw new Error('GitHub API response does not contain content or sha');
      }
      
      // Decode base64 content
      const content = Buffer.from(response.data.content, 'base64').toString('utf-8');
      const sha = response.data.sha;
      
      return [content, sha];
    } catch (error) {
      logger.error('Error fetching worklog content from GitHub:', error);
      throw error;
    }
  }
  
  /**
   * Update the worklog file with new content
   * @param newContent New content for the worklog file
   * @param sha SHA of the current file content
   * @returns True if update was successful
   */
  async updateWorklog(newContent: string, sha: string): Promise<boolean> {
    const { githubRepo, githubWorklogPath, logger } = this.config;
    
    // Split repo into owner and repo
    const [owner, repo] = githubRepo.split('/');
    
    if (!owner || !repo) {
      throw new Error(`Invalid GitHub repository format: ${githubRepo}. Expected format: owner/repo`);
    }
    
    logger.info(`Updating worklog content in ${githubRepo}/${githubWorklogPath}`);
    
    try {
      await this.octokit.repos.createOrUpdateFileContents({
        owner,
        repo,
        path: githubWorklogPath,
        message: 'Update worklog with Toggl time entries',
        content: Buffer.from(newContent).toString('base64'),
        sha,
      });
      
      logger.info('Worklog updated successfully');
      return true;
    } catch (error) {
      logger.error('Error updating worklog content on GitHub:', error);
      throw error;
    }
  }
  
  /**
   * Find an existing entry for the given date in the worklog content
   * @param content Worklog content
   * @param date Date to find
   * @returns Entry details if found, null otherwise
   */
  findEntryForDate(content: string, date: Date): WorklogEntry | null {
    const { timezone, logger } = this.config;
    
    // Convert to timezone
    const zonedDate = utcToZonedTime(date, timezone);
    
    // Format date as YYYY-M-D
    const dateStr = `${zonedDate.getFullYear()}-${zonedDate.getMonth() + 1}-${zonedDate.getDate()}`;
    
    // Look for an entry starting with the date
    const pattern = new RegExp(`(${dateStr} [A-Za-z]+ \\(\\d+\\.?\\d*h\\+?\\): .+?)(\\n\\n|\\n\\*|\\Z)`, 's');
    const match = content.match(pattern);
    
    if (match && match.index !== undefined) {
      const entry = {
        content: match[1],
        startIndex: match.index,
        endIndex: match.index + match[1].length,
      };
      
      logger.info(`Found existing entry for ${dateStr}: ${entry.content.substring(0, 50)}...`);
      return entry;
    }
    
    logger.info(`No existing entry found for ${dateStr}`);
    return null;
  }
  
  /**
   * Update an existing entry or create a new one
   * @param content Current worklog content
   * @param newEntry New entry to add or update
   * @param date Date of the entry
   * @returns Updated worklog content
   */
  updateOrCreateEntry(content: string, newEntry: string, date: Date): string {
    const { timezone, logger } = this.config;
    const zonedDate = utcToZonedTime(date, timezone);
    const dateStr = format(zonedDate, 'yyyy-M-d');
    
    // Check if there's already an entry for the date
    const existingEntry = this.findEntryForDate(content, date);
    
    let updatedContent: string;
    
    if (existingEntry) {
      // Update existing entry
      updatedContent = (
        content.substring(0, existingEntry.startIndex) + 
        newEntry + 
        content.substring(existingEntry.endIndex)
      );
      logger.info(`Updated existing entry for ${dateStr}`);
    } else {
      // Add new entry at the top
      if (content.startsWith('# ')) {
        // If the file starts with a title, add after the title
        const titleEnd = content.indexOf('\n') + 1;
        updatedContent = (
          content.substring(0, titleEnd) + 
          '\n' + newEntry + '\n\n' + 
          content.substring(titleEnd)
        );
      } else {
        // Otherwise, add at the very top
        updatedContent = newEntry + '\n\n' + content;
      }
      logger.info(`Added new entry for ${dateStr}`);
    }
    
    return updatedContent;
  }
}

File: python/toggl_github_sync/api/github.py
"""GitHub API client implementation."""

import base64
import logging
import re
from datetime import datetime
from typing import Dict, List, Optional, Tuple, Any

import pytz
from github import Github, GithubException
from github.ContentFile import ContentFile
from github.Repository import Repository

from toggl_github_sync.config import Config

logger = logging.getLogger(__name__)

class GitHubApiClient:
    """Client for interacting with the GitHub API."""

    def __init__(self, config: Config):
        """Initialize the GitHub API client.
        
        Args:
            config: Application configuration
        """
        self.github = Github(config.github_token)
        self.repo_name = config.github_repo
        self.worklog_path = config.github_worklog_path
        self.timezone = pytz.timezone(config.timezone)
        
    def get_worklog_content(self) -> Tuple[str, str]:
        """Get the current content of the worklog file.
        
        Returns:
            Tuple of (content, sha)
        """
        logger.info(f"Fetching worklog content from {self.repo_name}/{self.worklog_path}")
        
        repo = self.github.get_repo(self.repo_name)
        try:
            file_content = repo.get_contents(self.worklog_path)
            if isinstance(file_content, List):
                raise ValueError(f"Expected a file, but {self.worklog_path} is a directory")
            
            content = base64.b64decode(file_content.content).decode('utf-8')
            return content, file_content.sha
        except GithubException as e:
            logger.error(f"Error getting worklog content: {e}")
            raise
    
    def update_worklog(self, new_content: str, sha: str) -> bool:
        """Update the worklog file with new content.
        
        Args:
            new_content: New content for the worklog file
            sha: SHA of the current file content
            
        Returns:
            True if update was successful
        """
        logger.info(f"Updating worklog content in {self.repo_name}/{self.worklog_path}")
        
        repo = self.github.get_repo(self.repo_name)
        try:
            repo.update_file(
                path=self.worklog_path,
                message="Update worklog with Toggl time entries",
                content=new_content,
                sha=sha,
            )
            logger.info("Worklog updated successfully")
            return True
        except GithubException as e:
            logger.error(f"Error updating worklog: {e}")
            raise
    
    def find_entry_for_date(self, content: str, date: datetime) -> Optional[Tuple[str, int, int]]:
        """Find an existing entry for the given date in the worklog content.
        
        Args:
            content: Worklog content
            date: Date to find
            
        Returns:
            Tuple of (entry, start_index, end_index) if found, None otherwise
        """
        # Format date as YYYY-M-D
        date_str = date.strftime("%Y-%-m-%-d")
        
        # Look for an entry starting with the date
        # More robust pattern that handles:
        # - Optional spaces after date
        # - Any day name (not just English)
        # - Various hour formats
        # - Multi-line descriptions
        pattern = rf"({date_str}\s+[A-Za-z]+\s*\(\d+\.?\d*h\+?\):\s*.+?)(?=\n\n|\n\*|\Z)"
        match = re.search(pattern, content, re.DOTALL | re.MULTILINE)
        
        if match:
            entry = match.group(1)
            start_index = match.start()
            end_index = match.end()
            logger.info(f"Found existing entry for {date_str}: {entry[:50]}...")
            return entry, start_index, end_index
        else:
            logger.warning(f"No existing entry found for {date_str}")
            return None
    
    def update_or_create_entry(self, content: str, new_entry: str, date: datetime) -> str:
        """Update an existing entry or create a new one.
        
        Args:
            content: Current worklog content
            new_entry: New entry to add or update
            date: Date of the entry
            
        Returns:
            Updated worklog content
        """
        # Check if there's already an entry for the date
        existing_entry = self.find_entry_for_date(content, date)
        
        if existing_entry:
            # Update existing entry
            entry, start_index, end_index = existing_entry
            updated_content = (
                content[:start_index] + 
                new_entry + 
                content[end_index:]
            )
            logger.info(f"Updated existing entry for {date.strftime('%Y-%-m-%-d')}")
        else:
            # Add new entry at the top
            if content.startswith("# "):
                # If the file starts with a title, add after the title
                title_end = content.find("\n") + 1
                updated_content = (
                    content[:title_end] + 
                    "\n" + new_entry + "\n\n" + 
                    content[title_end:]
                )
            else:
                # Otherwise, add at the very top
                updated_content = new_entry + "\n\n" + content
            logger.info(f"Added new entry for {date.strftime('%Y-%-m-%-d')}")
        
        return updated_content
